pageserver/src/layered_repository/timeline.rs

```diff
@@ -319,7 +319,22 @@ pub struct LayeredTimeline {
     repartition_threshold: u64,
 
     /// Current logical size of the "datadir", at the last LSN.
-    current_logical_size: AtomicIsize,
+    ///
+    /// Size shouldn't ever be negative, but this is signed for two reasons:
```

Comment:
Thanks for the comment. Would also be helpful to define logical size. It's not obvious at all, since logical/physical is a relative term. Even if it's well defined in postgres:
1. It doesn't come up with a google search
2. It's not clear if neon logical == postgres logical

Without the definition, I find it hard to understand:
1. Why would we need lazy loading of size? Is it that complicated to compute?
2. If calculating relation size is critical and equally error prone as calculating total logical size, then do we gain much from tolerating negative relation size bugs?

Reply:
Great questions! 

As far as I'm aware concepts of logical/physical sizes are our own. Logical size being the sum of the sizes for all relations and physical size being the total size that is occupied on the file system including non relational data, WAL, etc.

> Why would we need lazy loading of size? Is it that complicated to compute?

There are two parts of it. For previous storage version we generated way more layers and if I remember correctly we needed to actually look inside them more closely (load index for old bookfile format). This with addition of huge layer numbers resulted in layer map loading and size calculation times up to 15 minutes with 1TB database (if I remember correctly, @knizhnik did the tests). So the answer is that with large databases it can take quite a while to calculate everything. But with newer storage it should be better than it was. With small database of 4.4GB logical size and 17.63 GB physical it took less than a millisecond to load layer map and around 250 milliseconds to initialize logical size. (calculated using #2278) Currently logical size is initialized by looking up all the relations and summing up their size, and with on-demand download it can take quite some time, because it may trigger downloads from remote storage.

Reply:
> Would also be helpful to define logical size. It's not obvious at all, since logical/physical is a relative term. 

+1. There's a brief comment in `get_current_logical_size_non_incremental`, in `pgdatadir_mapping.rs`, but it's very brief and not easy to find. I opened https://github.com/neondatabase/neon/pull/2306 to add it to the glossary in the docs.